Adding new Dell EMC ECS Object Storage Plugin for CloudStack

mhkadhum opened the pull request:
### Description

This PR adds support for a Dell EMC ECS S3 object storage plugin for Apache CloudStack. ECS is a software-defined object storage platform that supports both object and file-system protocols, with a focus on scalable and reliable object storage. We have been using ECS in production for three years, and this work extends CloudStack’s existing MinIO plugin to provide similar functionality for ECS.

The implementation supports the full lifecycle of S3-compatible buckets on ECS, including user provisioning, bucket creation, policy management, versioning, creation-time encryption, and integration with CloudStack’s S3 Browser. All functionality has been tested in a lab environment on Ubuntu using a CloudStack development setup based on the official installation guidelines.

Key architectural differences from MinIO:

**Management API Integration**  
ECS requires use of the ECS Management API (port 4443, or 443 when fronted by HAProxy). CloudStack authenticates with management-user credentials to perform bucket and user operations.

**Namespace Requirements**  
A dedicated ECS namespace is required for CloudStack-managed buckets. Multiple namespaces allow different CloudStack environments to share the same ECS cluster.

**S3 Endpoints (Public and Private)**  
ECS exposes S3 services on ports 9020/9021 (non-TLS/TLS). In our deployment, these are routed through HAProxy and exposed externally on port 443. The Public URL is displayed to CloudStack users, while the Private URL is used internally.

**TLS Handling**  
The “Allow Insecure HTTPS” option controls whether CloudStack accepts untrusted certificates when communicating with the ECS Management API.

**User Provisioning Workflow**  
When a CloudStack user creates their first bucket, CloudStack provisions a corresponding ECS object user using the CloudStack UUID with a `cs-` prefix. ECS generates access and secret keys once, which CloudStack securely stores and reuses for subsequent bucket operations.

Bucket features:

- **Encryption:** Supported only during bucket creation; CloudStack hides encryption in the update view.
- **Bucket Policy:** Supports Private and Public configurations.
- **Versioning:** Fully supported through the S3 API rather than the Management API.
- **Object Lock:** Not supported in this release; the UI hides the option and API calls return a clear error.

Bucket modification supports quota changes, versioning updates, and policy changes. Encryption is excluded because ECS does not allow changing it after creation.

CloudStack’s S3 Browser supports upload, download, delete, listing, and prefix filtering through the ECS S3 endpoint. ECS prevents deletion of non-empty buckets, and CloudStack surfaces these errors accordingly.

We welcome review and feedback. The development fork is available here:  
https://github.com/mhkadhum/cloudstack

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)  
- [x] New feature (non-breaking change which adds functionality)  
- [ ] Bug fix (non-breaking change which fixes an issue)  
- [ ] Enhancement (improves an existing feature and functionality)  
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)  
- [ ] Build/CI  
- [ ] Test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [x] Major  
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER  
- [ ] Critical  
- [ ] Major  
- [ ] Minor  
- [ ] Trivial

### Screenshots (if appropriate):

<img width="722" height="743" alt="Screenshot from 2025-11-24 16-58-08" src="https://github.com/user-attachments/assets/0208acfe-6df8-46c6-9c89-3d32cceb9277" />

<img width="708" height="558" alt="image" src="https://github.com/user-attachments/assets/95598ca4-dc29-4a99-9163-c346342132a6" />

<img width="550" height="356" alt="image" src="https://github.com/user-attachments/assets/19cd98c1-2398-4cda-8832-be7fddbf024f" />


### How Has This Been Tested?

The plugin was tested in an Ubuntu-based CloudStack development environment created by cloning the CloudStack source and following the official installation instructions. Testing included:

- Management API authentication and namespace operations  
- Object-user provisioning and credential handling  
- Bucket creation, deletion, versioning, quota updates, and policy changes  
- S3 Browser operations (upload, download, delete, listing, prefix navigation)  
- Error handling for invalid configurations and unsupported Object Lock requests  
- TLS and insecure-TLS scenarios  
- HAProxy-fronted deployments for both Management API and S3 endpoints

Commit:
Adding New Dell ECS Plugin to the API & editing Frontend for the ECS Provider

mhkadhum committed:
Update AddObjectStorage.vue ECS Field Names and desc.

mhkadhum committed:
Merge branch 'apache:main' into main

jbampton reviewed (commented):
You have failing tests for `pre-commit` and `license checks`.

Have a quick read up on pre-commit it really is a series of basic checks we run:

https://github.com/apache/cloudstack/blob/main/PRE-COMMIT.md

Also we normally add the ASF license header to all files or add an exclude

DaanHoogland commented on plugins/storage/object/ECS/src/main/resources/META-INF/cloudstack/storage-object-ecs/spring-storage-object-ecs-context.xml:
```diff
@@ -0,0 +1,14 @@
+         for lifecycle and driver, same as Ceph. -->
+    <bean id="EcsObjectStoreProviderImpl"
+          class="org.apache.cloudstack.storage.datastore.provider.EcsObjectStoreProviderImpl" />
+</beans>
```
```suggestion
</beans>

```

Commit:
add missing license

mhkadhum committed:
reverting update by pre-commit on tools/marvin/setup.py

DaanHoogland commented:
looks generally good @mhkadhum , thanks for the contribution. Will you be maintaining this? As it is a 3rd party component and it may require hardware or licenses not available to the project.

mhkadhum committed:
adding ECS Module in the Plugins pom.xml

DaanHoogland commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/EcsObjectStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,255 @@
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public DataStore initialize(Map<String, Object> dsInfos) {
```
this method can be split into several verify methods. Your code to maintain @mhkadhum , so I won’t -1 on this point but for readability it would be good.

DaanHoogland reviewed (commented):
I have not found any oddities in the code, but one remark (in two parts ;) ) Feel free to ignore but I think it will help in the future.

Commit:
Update ECS Driver, split to small helpers , remove SuppressWarnings from lifecycle , introduce caching to ECS token

mhkadhum commented:
> > Hello @abh1sar @DaanHoogland I just wanted to quickly follow up on my last comment. I understand u r probably busy but if u get a chance, could u take a look when possible ? Appreciate the help!
> 
> sorry @mhkadhum , this [#12124 (comment)](https://github.com/apache/cloudstack/pull/12124#issuecomment-3606486217) you mean right? I didn’t read it as a question and though this is ready for testing. What issues are you having right now?

No worries @DaanHoogland 
Yes this one. There are no issues at the moment. I am simply checking the status of the PR. Additionally, I am planning to develop another plugin for a different third-party object storage.
Do you have an ECS environment where you plan to conduct the testing?

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+        return d == null ? null : d.getValue();
+    }
+
+    private static Integer safeIntFromGetter(final Object o, final String getter) {
```
At present, this method is only invoked for Bucket.getQuota and Bucket.getSize (returning Integer and Long).
It looks like it is not reaquired at the moment and adds some additional complexity compared to handling these cases directly.

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+        final Map<String, String> ds = storeDetailsDao.getDetails(storeId);
+        final EcsCfg cfg = ecsCfgFromDetails(ds, storeId);
+
+        final String username = "cs-" + acct.getUuid();
```
Define "cs-" as a constant variable along with other constants

mhkadhum replied:
This feature is done and i also updated the FE.
<img width="1413" height="946" alt="image" src="https://github.com/user-attachments/assets/0242556c-2917-416d-8b3d-c27a0348d033" />

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+                final List<String> names = extractAllTags(body, "Name");
+                for (String n : names) {
+                    if (isBlank(n)) continue;
+                    final org.apache.cloudstack.storage.object.Bucket b =
```
can we just do Bucket b = new BucketObject();?

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+    }
+
+    @Override public AccessControlList getBucketAcl(final BucketTO bucket, final long storeId) { return null; }
+    @Override public void setBucketAcl(final BucketTO bucket, final AccessControlList acl, final long storeId) { /* not supported */ }
```
```suggestion
    @Override
    public void setBucketAcl(final BucketTO bucket, final AccessControlList acl, final long storeId) {
        /* not supported */
    }
```

mhkadhum replied:
Ok

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+
+        // Encryption flag from request/VO best-effort
+        boolean encryptionEnabled =
+            getBooleanFlagLoose(bucket, "getEncryption", "isEncryption", false) ||
```
The Bucket class has only isEncryption method which return a boolean.
The method getBooleanFlagLoose method seems like an overkill. we can just do `boolean encryptionEnabled = bucket.isEncryption()`.

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+    private BucketVO resolveBucketVO(final BucketTO bucket, final long storeId) {
+        long id = getLongFromGetter(bucket, "getId", -1L);
+        if (id <= 0) id = getLongFromGetter(bucket, "getBucketId", -1L);
+        if (id > 0) {
```
The bucket's id is defined as NOT NULL AUTO_INCREMENT in the schema. So, it can never be <= 0.

The code below line 1274 is not required. I don't think it will ever happen that bucketVO is not found using the id , but it is found using uuid or name.

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+                }
+            }
+
+            String accessKey = vo != null ? safeString(vo, "getAccessKey") : null;
```
I don't think we need reflections for getters anywhere in this file. 
The interfaces are clearly defined.
this can be written as 
 String accessKey = vo.getAccessKey()

abh1sar reviewed (commented):
Reviewed EcsObjectStoreDriverImpl
Overall functionality looks good but there are some comments on code maintainability and organisation. Please check.
Overall, I think use of reflections can be avoided in most cases.

mhkadhum commented:
> Reviewed EcsObjectStoreDriverImpl Overall functionality looks good but there are some comments on code maintainability and organisation. Please check. Overall, I think use of reflections can be avoided in most cases.

Hello @abh1sar 
Thank you for ur feedback, am working on addressing all the issues u provided.

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+        final String name  = bucket.getName();
+
+        if (objectLock) {
+            throw new CloudRuntimeException("Dell ECS doesn't support this feature: object locking");
```
ive changed this to InvalidParameterValueException

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+    private static final long EXPIRY_SKEW_SEC = 30;           // refresh early
+    private static final ConcurrentHashMap<TokenKey, TokenEntry> TOKEN_CACHE = new ConcurrentHashMap<>();
+    private static final ConcurrentHashMap<TokenKey, Object> TOKEN_LOCKS = new ConcurrentHashMap<>();
+
```
I agree.
Token management and XML parsing have been extracted into separate classes
(EcsMgmtTokenManager and EcsXmlParser) to improve maintainability and reuse.

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+import com.cloud.utils.exception.CloudRuntimeException;
+
+public class EcsObjectStoreDriverImpl extends BaseObjectStoreDriverImpl {
+    private static final Logger logger = LogManager.getLogger(EcsObjectStoreDriverImpl.class);
```
Deleted.

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+        return null;
+    }
+
+    private static boolean isBlank(final String s) {
```
Fixed

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+    private static final String SA_PASS    = "sa_password";   // service account password
+    private static final String NAMESPACE  = "namespace";     // e.g. cloudstack
+    private static final String INSECURE   = "insecure";      // "true" to ignore TLS cert/host
+    private static final String S3_HOST    = "s3_host";       // S3 endpoint host (or URL if UI provides it)
```
Done.

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+            if (vo != null) {
+                vo.setBucketURL(s3UrlForUI);
+                if (!isBlank(accKey)) vo.setAccessKey(accKey);
+                if (!isBlank(secKey)) vo.setSecretKey(secKey);
```
Done.

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1556 @@
+        Integer quotaGb = null;
+        try {
+            quotaGb = safeIntFromGetter(bucket, "getQuota");
+            if (quotaGb == null) quotaGb = safeIntFromGetter(bucket, "getSize");
```
Ok

mhkadhum committed:
Merge branch 'apache:main' into main

Commit:
Update ECS Plugin: address reviewer comments for ECS Plugin

DaanHoogland commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsMgmtTokenManager.java:
```diff
@@ -0,0 +1,142 @@
+package org.apache.cloudstack.storage.datastore.driver;
```
```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.cloudstack.storage.datastore.driver;
```

DaanHoogland commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsXmlParser.java:
```diff
@@ -0,0 +1,71 @@
+package org.apache.cloudstack.storage.datastore.driver;
```
```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.cloudstack.storage.datastore.driver;
```

mhkadhum committed:
adding license to ecs new files

mhkadhum committed:
adding license to ecs new files

mhkadhum committed:
adding license to ecs new files

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with no SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9  ✖️ debian ✖️ suse15. SL-JID 16184

mhkadhum committed:
fixing spaces

mhkadhum committed:
Update EcsObjectStoreDriverImpl.java

mhkadhum committed:
Update EcsObjectStoreDriverImpl.java

mhkadhum commented:
Hello @blueorangutan 
Had some stuff going on my build now should work now

from my server
mvn -pl :cloud-plugin-storage-object-ecs -Dcheckstyle.consoleOutput=true checkstyle:check
[INFO] Scanning for projects...
[INFO] 
[INFO] -------< org.apache.cloudstack:cloud-plugin-storage-object-ecs >--------
[INFO] Building Apache CloudStack Plugin - ECS object storage provider 4.23.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.0:check (default-cli) @ cloud-plugin-storage-object-ecs ---
[INFO] Starting audit...
Audit done.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.275 s
[INFO] Finished at: 2025-12-29T13:25:30+03:00
[INFO] ------------------------------------------------------------------------

abh1sar commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -165,26 +240,66 @@ export default {
-          name: values.name,
-          size: values.size
+        const data = {
+          name: values.name
```
size can apply to all providers (it's for framework level tracking)

mhkadhum replied:
ok

abh1sar commented on ui/src/views/storage/UpdateBucket.vue:
```diff
@@ -110,36 +132,31 @@ export default {
-        const values = this.handleRemoveFields(formRaw)
-
-        var data = {
+        const values = toRaw(this.form)
```
Why remove `const values = this.handleRemoveFields(formRaw)`?

DaanHoogland commented:
sorry @mhkadhum , still compile errors (not lint anymore): https://github.com/apache/cloudstack/actions/runs/20570652663/job/59082786785?pr=12124#step:7:18860

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1560 @@
+public class EcsObjectStoreDriverImpl extends BaseObjectStoreDriverImpl {
+
+    // ---- Injected dependencies ----
+    @Inject private AccountDao accountDao;
```
Fixed

mhkadhum commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -165,26 +240,66 @@ export default {
           data['details[3].value'] = values.s3Url
           data['details[4].key'] = 'iamUrl'
           data['details[4].value'] = values.iamUrl
+        } else if (provider === 'ECS') {
```
Added

mhkadhum commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,82 @@
+          <!-- S3 URL (for UI: this becomes addObjectStoragePool url=...) -->
+          <a-form-item name="url" ref="url">
+            <template #label>
+              <tooltip-label :title="'ECS Public URL'" :tooltip="'The S3-compatible endpoint URL that clients use to connect to ECS'"/>
```
done

mhkadhum commented on ui/src/views/storage/CreateBucket.vue:
```diff
@@ -110,18 +118,56 @@ export default {
+      }
+
+      const provider = (
+        selected.objectstoreprovider ||
```
my code was appearing to work before because it relied on fallbacks objectstoreprovider, provider, and even the object store name containing ECS However that behavior was not correct
it updated now and uses providername

Commit:
Update versioning logic, adress the reviwer comment for the UI

Commit:
add missing lice.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with no SystemVM templates. I'll keep you posted as I make progress.

abh1sar commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -137,68 +93,88 @@ public DataStoreTO getStoreTO(final DataStore store) {
-                getBooleanFlagLoose(vo, "getEncryptionEnabled", "isEncryptionEnabled", false);
+        // Fallback to persisted value if request did not explicitly enable it.
+        if (!encryptionEnabled) {
+            try {
```
this try is also not required

mhkadhum commented:
> @mhkadhum , it looks like you are ready to merge. If you wish to ignore my latest comments please let us know. (cc @abh1sar ).

Lets go ahead with the merge 
Thank you

abh1sar commented:
> > @mhkadhum , it looks like you are ready to merge. If you wish to ignore my latest comments please let us know. (cc @abh1sar ).
> 
> Lets go ahead with the merge 
> Thank you 

I am going to do some regression testing. Should be finished by tomorrow.

abh1sar commented:
> > > @mhkadhum , it looks like you are ready to merge. If you wish to ignore my latest comments please let us know. (cc @abh1sar ).
> > 
> > 
> > Lets go ahead with the merge
> > Thank you
> 
> I am going to do some regression testing. Should be finished by tomorrow.

Tested common object storage workflows with Ceph.
Add/Delete Object Store. Create/Update/Delete Bucket. Working as expected.

@mhkadhum Please answer any pending comments and we can move forward with this.

abh1sar reviewed (approved):
LGTM

mhkadhum committed:
remove no-needed try blocks

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -137,68 +93,88 @@ public DataStoreTO getStoreTO(final DataStore store) {
-        // Quota from UI (INT GB)
+        // Quota from UI (INT GB). Bucket.getQuota may be Integer; Bucket.getSize may be Long.
         Integer quotaGb = null;
         try {
```
https://github.com/mhkadhum/cloudstack/commit/6f3efbefc9ae8ba2ab62e00e1e0df472fd5beca0

mhkadhum commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsConstants.java:
```diff
@@ -0,0 +1,38 @@
+package org.apache.cloudstack.storage.datastore.driver;
+
+public final class EcsConstants {
+    private EcsConstants() {}
```
I am keeping this just to follow the standard pattern.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with no SystemVM templates. I'll keep you posted as I make progress.

mhkadhum committed:
Merge branch 'apache:main' into main

Commit:
running pre-commit

mhkadhum commented:
> @mhkadhum please fix build errors `EcsObjectStoreDriverImpl.java:120: Line has trailing spaces. [RegexpSingleline]`

Missed the pre-commit hook. Fixed and pushed.

abh1sar commented:
@blueorangutan package

blueorangutan commented:
@abh1sar a [SL] Jenkins job has been kicked to build packages. It will be bundled with no SystemVM templates. I'll keep you posted as I make progress.

mhkadhum committed:
Merge branch 'main' into main

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with no SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9  ✔️ debian ✖️ suse15. SL-JID 16379

blueorangutan commented:
@mhkadhum a [SL] Jenkins job has been kicked to build packages. It will be bundled withHello @DaanHoogland ,
Can u test, I believe this is ready to merge. SystemVM template(s). I'll keep you posted as I make progress.

winterhazel commented:
@mhkadhum thanks for the pull request, and sorry for the lack of attention to it recently. As @abh1sar already validated that there were no regression for other plugins in https://github.com/apache/cloudstack/pull/12124#issuecomment-3748034686, I will merge it to 4.23 with one more approval. The test failures in https://github.com/apache/cloudstack/pull/12124#issuecomment-3758633939 are unrelated.

I intend to review it this week.

Also, as you will be maintaining this plugin, you can add a line to `.github/CODEOWNERS` if you want to containing your GitHub handle so that you are automatically marked as a reviewer whenever a PR that affects the plugin is created.

mhkadhum committed:
Merge branch 'apache:main' into main

mhkadhum committed:
adding mhkadhum as maintainer for the ECS Storage Plugin

mhkadhum commented:
> @mhkadhum thanks for the pull request, and sorry for the lack of attention to it recently. As @abh1sar already validated that there were no regression for other plugins in [#12124 (comment)](https://github.com/apache/cloudstack/pull/12124#issuecomment-3748034686), I will merge it to 4.23 with one more approval. The test failures in [#12124 (comment)](https://github.com/apache/cloudstack/pull/12124#issuecomment-3758633939) are unrelated.
> 
> I intend to review it this week.
> 
> Also, as you will be maintaining this plugin, you can add a line to `.github/CODEOWNERS` if you want to containing your GitHub handle so that you are automatically marked as a reviewer whenever a PR that affects the plugin is created.

Hi @winterhazel 

I have synced the fork with its upstream source and registered myself within the CODEOWNERS file

winterhazel commented:
@blueorangutan package

blueorangutan commented:
@winterhazel a [SL] Jenkins job has been kicked to build packages. It will be bundled with no SystemVM templates. I'll keep you posted as I make progress.

winterhazel commented on .github/CODEOWNERS:
```diff
@@ -18,6 +18,7 @@
 /plugins/storage/volume/linstor @rp-
 /plugins/storage/volume/storpool @slavkap
 /plugins/storage/volume/ontap @rajiv1 @sandeeplocharla @piyush5 @suryag
+/plugins/storage/object/ECS @mhkadhum
```
Other directories use a lowercase name. Could you rename the plugin's directory to standardize it?

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ el10 ✔️ debian ✔️ suse15. SL-JID 17600

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/EcsObjectStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,321 @@
+    private static String getString(final Map<String, Object> dsInfos, final String key, final boolean required) {
+        final Object v = dsInfos.get(key);
+        final String s = (v == null) ? null : v.toString().trim();
+        if (required && (s == null || s.isEmpty())) {
```
```suggestion
        if (required && StringUtils.isEmpty(s)) {
```

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/EcsObjectStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,321 @@
+            return Long.parseLong(v.toString().trim());
+        } catch (Exception e) {
+            throw new CloudRuntimeException("ECS: invalid long for '" + key + "': " + v);
+        }
```
You can probably just cast `v` to a `Long`

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/EcsObjectStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,321 @@
+        final String url = getString(dsInfos, "url", true);
+        final String name = getString(dsInfos, "name", true);
+        final Long size = getLong(dsInfos, "size"); // optional
+        final String providerName = getProviderName(dsInfos);
```
You can use `PROVIDER_NAME` directly without having to obtain it from `dsInfos`. This method will not be called if `providerName` does not match `PROVIDER_NAME`.

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/EcsObjectStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,321 @@
+        }
+    }
+
+    private String loginAndGetToken(final String mgmtUrl, final String user, final String pass, final boolean insecure) {
```
The return value of this method is never used.

Also, this is very similar to `org.apache.cloudstack.storage.datastore.driver.EcsMgmtTokenManager#loginAndGetTokenFresh`. Perharps it could be deduplicated to make maintenance easier?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/EcsObjectStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,321 @@
+        return (m == null || m.isEmpty()) ? t.getClass().getSimpleName() : m;
+    }
+
+    private CloseableHttpClient buildHttpClient(final boolean insecure) {
```
This method is duplicated in `org.apache.cloudstack.storage.datastore.driver.EcsObjectStoreDriverImpl#buildHttpClient`

Maybe it could be unified somewhere to make maintenance easier?

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+            <template #label>
+              <tooltip-label
+                :title="'ECS Public URL'"
+                :tooltip="'The S3-compatible endpoint URL that clients use to connect to ECS'"
```
Could you extract this text to the English translation file (`en.json`) and use its key instead?

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+            <template #label>
+              <tooltip-label
+                :title="'ECS API URL'"
+                :tooltip="'ECS management API URL'"
```
Extract the text

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+            <template #label>
+              <tooltip-label
+                :title="'ECS Private URL'"
+                :tooltip="'The internal S3 endpoint URL used by CloudStack to communicate with ECS. May be the same as the Public URL.'"
```
Extract the text

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+            <template #label>
+              <tooltip-label
+                :title="'ECS service account user'"
+                :tooltip="'Service account user, e.g. cloudstack'"
```
Extract the text

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+            <template #label>
+              <tooltip-label
+                :title="'ECS service account password'"
+                :tooltip="'Service account password (sa_password)'"
```
Extract the text

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+              <tooltip-label
+                :title="'Namespace'"
+                :tooltip="'ECS namespace (namespace), e.g. cloudstack'"
+              />
```
Extract the text

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+            <template #label>
+              <tooltip-label
+                :title="'User prefix'"
+                :tooltip="'Prefix used for ECS user creation. Default is cs- (user_prefix). Example: cs-'"
```
Extract the text

winterhazel commented on ui/src/views/infra/AddObjectStorage.vue:
```diff
@@ -67,16 +68,106 @@
+
+          <a-form-item name="insecure" ref="insecure">
+            <a-checkbox v-model:checked="form.insecure">
+              Allow insecure HTTPS (set insecure=true)
```
Extract the text

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+
+    // ---------------- helpers ----------------
+
+    static final class EcsCfg {
```
This class is very similar to the EcsConfig from the LifeCycle class. We could unify it

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        }
+
+        final Map<String, String> ds = storeDetailsDao.getDetails(storeId);
+        final EcsCfg cfg = ecsCfgFromDetails(ds, storeId);
```
This bit is repeated in many methods. Maybe have a `ecsCfgFromDetails` that only receives the `storeId`, obtains the details form the DB, and creates an `EcsCfg` with the additional parameters such as the user prefix?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        return d.getValue();
+    }
+
+    private static String trimTail(final String s) {
```
This one is identical to `org.apache.cloudstack.storage.datastore.lifecycle.EcsObjectStoreLifeCycleImpl#trim`. It could go into a Utils file, along with other methods that are not directly related to the driver/lifecycle, such as `valueOrNull`.

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        final String ownerUser = getUserPrefix(ds) + acct.getUuid();
+
+        // Ensure per-account credentials exist (single-key policy with adopt-if-exists)
+        ensureAccountUserAndSecret(accountId, ownerUser, cfg.mgmtUrl, cfg.saUser, cfg.saPass, cfg.ns, cfg.insecure);
```
Pass the `EcsCfg` instead?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsMgmtTokenManager.java:
```diff
@@ -0,0 +1,161 @@
+
+    static final class EcsUnauthorizedException extends RuntimeException {
+        EcsUnauthorizedException(final String msg) { super(msg); }
+    }
```
This one doesn't seem to be strict to `EcsMgmtTokenManager`. Extract it to a separate class? Also, it could extend `CloudRuntimeException` instead.

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        final EcsCfg cfg = new EcsCfg(trimTail(mgmtUrl), saUser, saPass, ns, insecure);
+
+        try {
+            tokenManager.callWithRetry401(cfg, token -> {
```
It would be nice to extract the inner code to a separate method

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                    }
+
+                    // If ACS already has key -> do NOT create another.
+                    if (!StringUtils.isBlank(haveAcc) && !StringUtils.isBlank(haveSec)) {
```
The code inside this if could go to a separate method

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                        logger.debug("Failed to fetch existing ECS keys for {} (proceeding to create one): {}", username, e.getMessage());
+                    }
+
+                    // No ECS key either -> create exactly ONE new secret and store in ACS
```
This could go into a separate `createNewSecret()` method

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                        return null;
+                    }
+
+                    // ACS does NOT have key -> try to ADOPT existing ECS key first
```
This could go into a separate method that returns `true` if an existing ECS secret was adopted, in order to return early here

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                        + "<is_encryption_enabled>" + (encryptionEnabled ? "true" : "false") + "</is_encryption_enabled>"
+                        + "</object_bucket_create>";
+
+        if (logger.isDebugEnabled()) {
```
This `if` is not neeed

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        }
+
+        try {
+            tokenManager.callWithRetry401(cfg, token -> {
```
The internal code could be extracted into a separate method

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        }
+    }
+
+    private S3Endpoint resolveS3Endpoint(final Map<String, String> ds, final long storeId) {
```
`storeId` is not used

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        final String saUser = ds.get(EcsConstants.SA_USER);
+        final String saPass = ds.get(EcsConstants.SA_PASS);
+        final String ns = StringUtils.isBlank(ds.get(EcsConstants.NAMESPACE)) ? "default" : ds.get(EcsConstants.NAMESPACE);
+        final boolean insecure = "true".equalsIgnoreCase(ds.getOrDefault(EcsConstants.INSECURE, "false"));
```
This assignment is repeated many times. Maybe turn it into a method?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        }
+
+        final boolean insecure = "true".equalsIgnoreCase(ds.getOrDefault(EcsConstants.INSECURE, "false"));
+        final java.util.List<Bucket> out = new java.util.ArrayList<>();
```
Import `java.util.ArrayList` and `java.util.List` to use only the class name

winterhazel commented on plugins/storage/object/ecs/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                .ofPattern("EEE, dd MMM yyyy HH:mm:ss 'GMT'", Locale.US)
+                .withZone(java.time.ZoneOffset.UTC);
+        return fmt.format(java.time.Instant.now());
+    }
```
Import the classes to use only the class name

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                new javax.crypto.spec.SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA1");
+        mac.init(sk);
+        final byte[] raw = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
+        return Base64.getEncoder().encodeToString(raw);
```
Import the classes to use only the class name

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+    }
+
+    // ---------------- S3: list objects in a bucket (SigV2, path-style) ----------------
+    public List<String> listBucketObjects(final String bucketName, final long storeId) {
```
This method is not used

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                final StringBuilder qs = new StringBuilder("max-keys=1000");
+                if (!StringUtils.isBlank(marker)) {
+                    qs.append("&marker=").append(java.net.URLEncoder.encode(
+                            marker, java.nio.charset.StandardCharsets.UTF_8.name()));
```
Import the classes

winterhazel commented on plugins/storage/object/ecs/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+    }
+
+    @Override
+    public AccessControlList getBucketAcl(final BucketTO bucket, final long storeId) {
+        return null;
+    }
+
+    @Override
+    public void setBucketAcl(final BucketTO bucket, final AccessControlList acl, final long storeId) {
+        // not supported
+    }
```
We could throw a `UnsupportedOperationException` for these unsupported operations

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        final String url;
+        try {
+            url = cfg.mgmtUrl + "/object/bucket/" + b + "/policy?namespace="
+                    + java.net.URLEncoder.encode(cfg.ns, java.nio.charset.StandardCharsets.UTF_8.name());
```
Import the classes

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        try {
+            tokenManager.callWithRetry401(cfg, token -> {
+                final String current = getBucketPolicyRaw(url, token, cfg.insecure); // "" if none
+                final boolean hasPolicy = current != null && !current.isBlank();
```
`current` is never null

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        final String url;
+        try {
+            url = cfg.mgmtUrl + "/object/bucket/" + bucketName + "/policy?namespace="
+                    + java.net.URLEncoder.encode(cfg.ns, java.nio.charset.StandardCharsets.UTF_8.name());
```
Import the classes

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+    public boolean setBucketEncryption(final BucketTO bucket, final long storeId) {
+        final String bucketName = bucket != null ? bucket.getName() : "<null>";
+        logger.info("ECS setBucketEncryption('{}'): ECS only supports encryption at bucket creation. Treating as no-op.", bucketName);
+        return true;
```
We could throw a `UnsupportedOperationException` to inform users that encryption is only supported for ECS at bucket creation

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                "Dell ECS bucket encryption can only be chosen at bucket creation; "
+                        + "it cannot be disabled afterwards (bucket=" + bucketName + ")";
+        logger.error("ECS deleteBucketEncryption('{}') requested but {}", bucketName, msg);
+        throw new CloudRuntimeException(msg);
```
Same as https://github.com/apache/cloudstack/pull/12124/changes#r3147549331, just replace the exception type

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+
+        if (ep == null || StringUtils.isBlank(ep.host)) {
+            logger.warn("ECS: S3 endpoint not resolvable; skipping bucket versioning.");
+            return true; // best-effort
```
Why return `true`? Doesn't it make more sense to fail the create/update operation informing users that it was not possible to enable/disable versioning instead of returning success on the creation of a bucket that does not match what they asked for?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+            if (vo != null) {
+                accountId = vo.getAccountId();
+            }
+        }
```
An admin may update the bucket of another user to enable versioning. Does it make sense to only obtain the accountId of the bucket's owner instead (to get their keys)?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+            return null;
+        }
+
+        final long id = getLongFromGetter(bucket, "getId", -1L);
```
`getId` does not exist for `BucketTO`. As this method is used to obtain the ID of the account that owns the bucket, why not use `bucket.getAccountId()`?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        } catch (Exception e) {
+            logger.warn("ECS: bucket versioning {} failed for '{}': {}",
+                    desired, bucketName, e.getMessage());
+            return true; // best-effort (do NOT break createBucket)
```
Same as https://github.com/apache/cloudstack/pull/12124/changes#r3147571138

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+    }
+
+    private static String base64Md5(final byte[] data) throws Exception {
+        final java.security.MessageDigest md = java.security.MessageDigest.getInstance("MD5");
```
Import the class

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+
+    @Override
+    public Map<String, Long> getAllBucketsUsage(final long storeId) {
+        throw new CloudRuntimeException("Bucket usage aggregation is not implemented via Mgmt API in this plugin.");
```
Throw an `UnsupportedOperationException` instead

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+                    }
+
+                    if (currentGb != null && size <= currentGb) {
+                        logger.info("ECS setBucketQuota noop for '{}': requested {}GB <= current {}GB", bucketName, size, currentGb);
```
Maybe throw a `CloudRuntimeException` informing that it was not possible to set the quota to the user-requested value?

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+    }
+
+    // ---------- Mgmt owner → accountId fallback ----------
+    private Long resolveAccountIdViaMgmt(final String bucketName, final Map<String, String> ds, final boolean insecure) {
```
This method is not used

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+        }
+    }
+
+    private String fetchBucketOwnerViaMgmt(final CloseableHttpClient http, final String mgmtUrl, final String token, final String bucketName) throws Exception {
```
This method will not be used after removing the previous one

winterhazel commented on plugins/storage/object/ECS/src/main/java/org/apache/cloudstack/storage/datastore/driver/EcsObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,1453 @@
+    }
+
+    // Check if a bucket exists on ECS via Mgmt API /object/bucket/{name}/info?namespace=...
+    private boolean ecsBucketExists(final String bucketName, final Map<String, String> ds) {
```
This method is not used

winterhazel commented on ui/src/views/storage/UpdateBucket.vue:
```diff
@@ -24,46 +24,67 @@
+          v-model:checked="form.objectlocking"
+          :checked="form.objectlocking"/>
       </a-form-item>
+
```
Parameter `objectlocking` is not available for the `updateBucket` API. Why was this introduced here?

winterhazel commented on ui/src/views/storage/UpdateBucket.vue:
```diff
@@ -80,26 +101,42 @@ export default {
+        r.objectStoreprovider ||
+        r.provider ||
+        ''
+      ).toString().toUpperCase()
```
I think `r.provider` is the only valid option

winterhazel commented on ui/src/views/storage/UpdateBucket.vue:
```diff
@@ -110,55 +147,59 @@ export default {
           this.$notification.error({
             message: `${this.$t('label.bucket.update')} ${this.$t('label.error')}`,
-            description: error.response.data.updatebucketresponse.errortext,
+            description: error.response?.data?.updatebucketresponse?.errortext || 'Error',
```
Extract the text

mhkadhum commented:
Hello @winterhazel 
Ill work on ur request this week and give u feedback

Commit:
refactor ecs plugin

mhkadhum committed:
Update CODEOWNERS from ECS to ecs

mhkadhum commented:
Rename plugin directory to lowercase ecs
Extract EcsCfg, EcsUtils, EcsUnauthorizedException as shared classes
Consolidate duplicated buildHttpClient and login logic into EcsUtils
Replace silent no-ops with UnsupportedOperationException for unsupported operations ACL  , getAllBucketsUsage , encryption removal
Fail setBucketQuota when ECS cannot reduce quota instead of silently ignoring the request
Remove unused methods: listBucketObjects, ecsBucketExists, resolveAccountIdViaMgmt
Switch setBucketVersioning to S3 SigV2 path-style
Extract all hardcoded ECS UI strings to en.json add user_prefix field to Add Object Storage form
Drop objectlocking from updateBucket payload simplify provider detection to r.provider

mhkadhum commented:
Hi @DaanHoogland @abh1sar  @winterhazel 

Can u run the pipeline i just want to check if there are any issues with the new plugin, the build and test went great in my lab env.

winterhazel commented:
@blueorangutan package

blueorangutan commented:
@winterhazel a [SL] Jenkins job has been kicked to build packages. It will be bundled with  no SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9  ✖️ debian ✖️ suse15. SL-JID 18127

winterhazel commented:
@mhkadhum there is a build error:

```
Error:  The build could not read 1 project -> [Help 1]
Error:    
Error:    The project org.apache.cloudstack:cloudstack-plugins:4.23.0.0-SNAPSHOT (/home/runner/work/cloudstack/cloudstack/plugins/pom.xml) has 1 error
Error:      Child module /home/runner/work/cloudstack/cloudstack/plugins/storage/object/ecs of /home/runner/work/cloudstack/cloudstack/plugins/pom.xml does not exist
```

Commit:
adding the new ecs plguin

Commit:
refactor ECS plugin

mhkadhum commented:
> @mhkadhum there is a build error:
> 
> ```
> Error:  The build could not read 1 project -> [Help 1]
> Error:    
> Error:    The project org.apache.cloudstack:cloudstack-plugins:4.23.0.0-SNAPSHOT (/home/runner/work/cloudstack/cloudstack/plugins/pom.xml) has 1 error
> Error:      Child module /home/runner/work/cloudstack/cloudstack/plugins/storage/object/ecs of /home/runner/work/cloudstack/cloudstack/plugins/pom.xml does not exist
> ```

Can u re-run the pipeline

winterhazel commented:
@blueorangutan package

blueorangutan commented:
@winterhazel a [SL] Jenkins job has been kicked to build packages. It will be bundled with  no SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ el10 ✔️ debian ✔️ suse15. SL-JID 18150

mhkadhum commented:
Hello @winterhazel 
Not sure why the Integration Tests Result Failed.
Did u review the new update?

winterhazel commented:
@mhkadhum I was not able to review the new changes yet. The test failure here in GitHub does not seem related to them. We can try running the integration tests with a real hypervisor to check.

winterhazel commented:
@DaanHoogland @weizhouapache @nvazquez can we run the CI here?

mhkadhum commented:
> @DaanHoogland @weizhouapache @nvazquez can we run the CI here?

Hi @DaanHoogland 
Can u check this please 🙂

winterhazel commented:
> > @DaanHoogland @weizhouapache @nvazquez can we run the CI here?
> 
> Hi @DaanHoogland Can u check this please 🙂

Also pinging @Pearl1594

Pearl1594 commented:
@blueorangutan test

blueorangutan commented:
@Pearl1594 a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

weizhouapache reviewed (approved):
java code lgtm

blueorangutan commented:
<b>[SF] Trillian test result (tid-16299)</b>
Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 49044 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr12124-t16299-kvm-ol8.zip
Smoke tests completed. 151 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

mhkadhum commented:
Hello @winterhazel @DaanHoogland @Pearl1594 

The test result looks good.

> [SF] Trillian test result (tid-16299) Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8 Total time taken: 49044 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr12124-t16299-kvm-ol8.zip Smoke tests completed. 151 look OK, 0 have errors, 0 did not run Only failed and skipped tests results shown below:
> 
> Test	Result	Time (s)	Test File

i am not sure how to resolve these two :/ 
<img width="908" height="144" alt="image" src="https://github.com/user-attachments/assets/6d20cac9-4076-4424-b27c-c7acb80081a0" />

mhkadhum commented:
> Hello @winterhazel @DaanHoogland @Pearl1594
> 
> The test result looks good.
> 
> > [SF] Trillian test result (tid-16299) Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8 Total time taken: 49044 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr12124-t16299-kvm-ol8.zip Smoke tests completed. 151 look OK, 0 have errors, 0 did not run Only failed and skipped tests results shown below:
> > Test	Result	Time (s)	Test File
> 
> i am not sure how to resolve these two :/ <img alt="image" width="908" height="144" src="https://private-user-images.githubusercontent.com/96632094/608368571-6d20cac9-4076-4424-b27c-c7acb80081a0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODE2NzczMjQsIm5iZiI6MTc4MTY3NzAyNCwicGF0aCI6Ii85NjYzMjA5NC82MDgzNjg1NzEtNmQyMGNhYzktNDA3Ni00NDI0LWIyN2MtYzdhY2I4MDA4MWEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjE3VDA2MTcwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2MzE0MzA5NjBkOTRlOTNmN2MyNTBlZGUzMTEyNzlmZTYwM2NjMjMzMzk3OTFlNDVmZWFmOWEyNmYzNjQxODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.fOEau1vPEZ8ClwxumAeNgqhVW46I0Q0kkNEiPWUPUF0">

@DaanHoogland

winterhazel commented:
> Hello @winterhazel @DaanHoogland @Pearl1594
> 
> The test result looks good.
> 
> > [SF] Trillian test result (tid-16299) Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8 Total time taken: 49044 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr12124-t16299-kvm-ol8.zip Smoke tests completed. 151 look OK, 0 have errors, 0 did not run Only failed and skipped tests results shown below:
> > Test	Result	Time (s)	Test File
> 
> i am not sure how to resolve these two :/ <img alt="image" width="908" height="144" src="https://private-user-images.githubusercontent.com/96632094/608368571-6d20cac9-4076-4424-b27c-c7acb80081a0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODE3MDI0NjMsIm5iZiI6MTc4MTcwMjE2MywicGF0aCI6Ii85NjYzMjA5NC82MDgzNjg1NzEtNmQyMGNhYzktNDA3Ni00NDI0LWIyN2MtYzdhY2I4MDA4MWEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjE3VDEzMTYwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyNjc5MmQ2OTcyZjZlMGQyYTcwNmNhYWZjYWZhMzljZmM2Mzk5OTZkZGJkZjMwNmY2MjNlNTdiNjBkODcxZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.9cy436sK7OCq-DwM3-uH4ZmBOCWHVHdfjyCzL11JDGs">

@mhkadhum the first failure is caused by the lack of unit tests (coverage under 18%) for these changes. Passing this check is not a requirement for merging, but of course it would be good to have tests to prevent future regressions, if you would like to add some. The second one does not seem related to your changes, but I triggered a rerun.

winterhazel commented:
@mhkadhum the failure is an issue with Codecov, not with your changes:

```
==> Verifying GPG signature integrity
 -> Downloading https://cli.codecov.io/latest/linux/codecov.SHA256SUM
 -> Downloading https://cli.codecov.io/latest/linux/codecov.SHA256SUM.sig
 
gpg: Signature made Tue Apr 21 19:28:03 2026 UTC
gpg:                using RSA key 27034E7FDB850E0BBC2C62FF806BB28AED779869
gpg: Can't check signature: No public key
==> Could not verify signature. Please contact Codecov if problem continues
    Exiting...
Error: Process completed with exit code 1.
```

mhkadhum commented:
@winterhazel  
> @mhkadhum the failure is an issue with Codecov, not with your changes:
> 
> ```
> ==> Verifying GPG signature integrity
>  -> Downloading https://cli.codecov.io/latest/linux/codecov.SHA256SUM
>  -> Downloading https://cli.codecov.io/latest/linux/codecov.SHA256SUM.sig
>  
> gpg: Signature made Tue Apr 21 19:28:03 2026 UTC
> gpg:                using RSA key 27034E7FDB850E0BBC2C62FF806BB28AED779869
> gpg: Can't check signature: No public key
> ==> Could not verify signature. Please contact Codecov if problem continues
>     Exiting...
> Error: Process completed with exit code 1.
> ```

Great, I think we are ready to merge ? 
regarding the unit test they can be added later as an improvement

winterhazel commented:
> @winterhazel
> 
> > @mhkadhum the failure is an issue with Codecov, not with your changes:
> > ```
> > ==> Verifying GPG signature integrity
> >  -> Downloading https://cli.codecov.io/latest/linux/codecov.SHA256SUM
> >  -> Downloading https://cli.codecov.io/latest/linux/codecov.SHA256SUM.sig
> >  
> > gpg: Signature made Tue Apr 21 19:28:03 2026 UTC
> > gpg:                using RSA key 27034E7FDB850E0BBC2C62FF806BB28AED779869
> > gpg: Can't check signature: No public key
> > ==> Could not verify signature. Please contact Codecov if problem continues
> >     Exiting...
> > Error: Process completed with exit code 1.
> > ```
> 
> Great, I think we are ready to merge ? regarding the unit test they can be added later as an improvement

@mhkadhum I think there were some issues while adding your changes on the latest commits. Some unrelated labels were modified or removed from `en.json`. Also, it seems that some of @abh1sar's suggestions were reverted as well. Could you have a look?

weizhouapache commented:
@mhkadhum 
could you fix the conflict ?

weizhouapache commented:
@abh1sar 
can you review it ? thanks

mhkadhum commented:
> @abh1sar can you review it ? thanks

Hello

I'm quite busy with other commitments this week and won't be able to make progress on this until next week.

thank u

DaanHoogland commented on ui/public/locales/en.json:
I am pretty sure that the removals in this file are invalid. The additions seem ecs related, but the removals are not.

weizhouapache commented:
moving to 4.24.0
